src/Workspaces/Core/Portable/CodeRefactorings/FixAllOccurences/FixAllState.cs

```diff
@@ -0,0 +1,135 @@
+            CodeRefactoringProvider codeRefactoringProvider,
+            FixAllScope fixAllScope,
+            string codeActionTitle,
+            string? codeActionEquivalenceKey)
+            : base(fixAllProvider, document, project, codeRefactoringProvider, fixAllScope, codeActionEquivalenceKey)
+        {
+            this.SelectionSpan = selectionSpan;
+            this.CodeActionTitle = codeActionTitle;
+        }
+
+        protected override FixAllState With(Document? document, Project project, FixAllScope scope, string? codeActionEquivalenceKey)
+        {
+            return new FixAllState(
+                this.FixAllProvider,
+                document,
+                project,
+                this.SelectionSpan,
+                this.Provider,
+                scope,
+                this.CodeActionTitle,
+                codeActionEquivalenceKey);
+        }
+
+        /// <summary>
+        /// Gets the spans to fix by document for the <see cref="FixAllScope"/> for this fix all occurences fix.
+        /// </summary>
+        internal async Task<ImmutableDictionary<Document, ImmutableArray<TextSpan>>> GetFixAllSpansAsync(CancellationToken cancellationToken)
+        {
+            IEnumerable<Document>? documentsToFix = null;
+            switch (this.Scope)
+            {
+                case FixAllScope.ContainingType or FixAllScope.ContainingMember:
+                    Contract.ThrowIfNull(Document);
+                    var spanMappingService = Document.GetLanguageService<IFixAllSpanMappingService>();
+                    if (spanMappingService is null)
+                        return ImmutableDictionary<Document, ImmutableArray<TextSpan>>.Empty;
+
+                    return await spanMappingService.GetFixAllSpansAsync(
+                        Document, SelectionSpan, Scope, cancellationToken).ConfigureAwait(false);
+
+                case FixAllScope.Document:
+                    Contract.ThrowIfNull(Document);
+                    documentsToFix = SpecializedCollections.SingletonEnumerable(Document);
+                    break;
+
+                case FixAllScope.Project:
+                    documentsToFix = Project.Documents;
+                    break;
+
+                case FixAllScope.Solution:
+                    documentsToFix = Project.Solution.Projects.SelectMany(p => p.Documents);
+                    break;
+
+                default:
+                    return ImmutableDictionary<Document, ImmutableArray<TextSpan>>.Empty;
+            }
+
+            using var _ = PooledDictionary<Document, ImmutableArray<TextSpan>>.GetInstance(out var builder);
+            foreach (var document in documentsToFix)
+            {
+                TextSpan span;
+                if (document.SupportsSyntaxTree)
+                {
+                    var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                    span = root.FullSpan;
```

Comment:
i don't get this, and it also seems enormously expensive.  First, you're parsing the entire file, just to get teh span of hte filer.  that can be determined purely with `GetTextAsync` and getting the length of htat.

Second, if 'the full span of the document' is important, it would be great to have a way to pass that that doesn't require reading in all files from disk up front.

Third, in the else block you pass 'default'.  what does that mean?  The entire file?  Or literally teh 0 span on position 0 of the file?

Reply:
I have tweaked this slightly - this API now avoids computing the full document span, and instead returns an empty array of spans for cases where entire document needs to be fixed. The place where we actually invoke FixAllAsync on each refactoring with each document handles the empty array case and computes the full span at that time.